pkgs: add consideration for new packages (#250352)
NixOS/nixpkgs: pkgs/README.md

## Quick Start to Adding a Package

Each new package has a cost for the maintainers, Hydra and users downloading
Nixpkgs. Before adding a new package, please consider the following questions:

> * Does the package have a release?

Author: @7c6f434c what better metric for maturity do you propose? Let's restart the convo in this thread.

Author: Also what kind of packages do you have in mind? The vast majority of packages have releases.

Reviewer: I don't think there is a good universal metric, and I don't think there can be. 

If what we care about is whether the package ready for general use, we should start by asking «Is the package ready for general use?» Sure, nothing wrong with adding «For example, if the author(s) is/are working towards the first release, it is probably too early.»

Quicklisp for one example is half-full of packages that don't bother with versions and just publish commits fixing whatever small bugs are discovered.

Author: Thanks, that's useful. Adopted.

Reviewer: Thanks!

* How realistic is it that it will be used by other people?

> * Are you going to maintain the package?

Reviewer 2: Wondering if we should specify a timeframe here, like

```suggestion
* Are you going to keep maintaining the package at least 6 months from today?
```

Author: It's a good question, I don't know.

On the surface, it seems reasonable. But it also opens up new questions like what is the appropriate duration, is there a single duration that works for all packages, and what happens if they don't respect that commitment?

Or we ask the contributor to make a good-faith commitment and leave it unspecified.

Reviewer 2: A more reasonable timeframe would be the next NixOS release cycle, but I'm fine with a good faith commitment as well.

Author: It would be nice to move towards that for sure. I wonder what existing maintainers would think of this. Keeping this open for now to let it simmer a bit.

Author: Added some wording to make the minimum duration normative but not enforced: a0378d09042d3cc4ed5166d20c7f0f9f27913511

## Quick Start to Adding a Package

> Each new package has a cost for the maintainers, Hydra and users downloading

Reviewer 3: ```suggestion
Each new package has a cost for the maintainers, Continuous Integration, caching servers and users downloading
```

More generic and user'ish.

Nixpkgs. Before adding a new package, please consider the following questions:

> * Is the package ready for general use? For example, if the author(s) is/are working towards the first release, it is probably too early.

Reviewer 4: Remark: I will go further, I had discussions with @Foxboron on this who had some of his projects "too early packaged" and he was not really comfortable with this.

For example, we can also be explicit and say "if the author explicitly asks for not being packaged", etc. I mean, it's obvious in retrospective in packaging `unstable-XYZ`-type of packages.

I would just like to distinguish:

- software that is used and can be used because it has been there for a while, but no release or security fixes in latest HEAD, etc.
- software that is being actively developed and can be already ~usable as a sneak peek, but with absolute no quality guarantees.

The former is fine without "first release", the latter is very much meh IMHO.

Reviewer: The precedent is that we don't do much deference to upstream, though. (As in, multipled upstreams tried to stop people from packaging their announced-as-stable OSI-and-DSFG-approved-license software, and Nixpkgs ignored the requests)

Reviewer 5: There is a difference between a maintainer going "this isn't stable software and I would urge you to wait" vs "I dont want my FOSS code published in a software repository".

I would ignore the latter if I didn't care about downstream support when encountering bugs.

Author: Ok, I addressed this in d5f8b87b2ece652c722a0091c793465e005f6406 . I tried compressing the decision tree so it doesn't become a full D&D adventure.

* Does the project have a license?
* How realistic is it that it will be used by other people?
* Are you going to maintain the package?

> If any of these questions' answer is no, then you should probably not add the

Reviewer 4: Offer alternatives here IMHO: "consider a NUR, etc."

Author: fixed in 7fae21eb99c81c2f3998db54509c6c81f6063003

## Quick Start to Adding a Package

We welcome new contributors of new packages to Nixpkgs, arguably the greatest software database known. However, each new package comes with a cost for the maintainers, Continuous Integration, caching servers and users downloading Nixpkgs.

Before adding a new package, please consider the following questions:

> * Is the package ready for general use? We don't want to include projects that are going to be abandoned immediately. In case of doubt, check with upstream.

Reviewer 3: That being said, a better wording:

```suggestion
* Is the package ready for general use? We don't want to include projects that are too imature or are going to be abandoned immediately. In case of doubt, check with upstream.
```

* Does the project have a clear license statement? Remember that softwares are unfree by default (all rights reserved), and merely providing access to the source code does not imply its redistribution. In case of doubt, ask upstream.

> * How realistic is it that it will be used by other people? It's good that nixpkgs caters to various niches, but if it's a niche of 5 people it's probably too small.

Reviewer 6: The list seems to be missing a bullet point about novelty programs and programs with multiple implementations in nixpkgs already

See https://github.com/NixOS/nixpkgs/pull/258479 which adds a 4th implementation of lolcat

## Quick Start to Adding a Package

> Each new package has a cost for the maintainers, Continuous Integration, caching servers and users downloading Nixpkgs.

Reviewer 3: ```suggestion
We welcome new contributors of new packages to Nixpkgs, arguably the greatest software database known. However, each new package comes with a cost for the maintainers, Continuous Integration, caching servers and users downloading Nixpkgs.
```

Before adding a new package, please consider the following questions:

> * Is the package ready for general use? We don't want to include projects that are going to be abandoned immediately. In case of doubt, check with upstream.

Author: These points seem close enough that they can be handled by that single point. Let's keep the list as short as possible.

* Does the project have a clear license statement? Remember that softwares are unfree by default (all rights reserved), and merely providing access to the source code does not imply its redistribution. In case of doubt, ask upstream.
* How realistic is it that it will be used by other people? It's good that nixpkgs caters to various niches, but if it's a niche of 5 people it's probably too small.
* Are you willing to maintain the package? You should care enough about the package to be willing to keep it up and running for at least one complete Nixpkgs' release life-cycle.

> If any of these questions' answer is no, then you should probably not add the
> package.

Reviewer 7: ```suggestion
If any of these questions' answer is no, then you should probably not add the package.
```
big fan of one sentence per line like the rest of the file :)

This is section describes a general framework of understanding and exceptions might apply.

> Luckily it's pretty easy to maintain your own package set with Nix , which can then be added to the [Nix User Repository](https://github.com/nix-community/nur) project.

Reviewer 7: ```suggestion
Luckily it's pretty easy to maintain your own package set with Nix, which can then be added to the [Nix User Repository](https://github.com/nix-community/nur) project.
```

   Some notes:

   - Add yourself as the maintainer of the package.

>    - All other [`meta`](https://nixos.org/manual/nixpkgs/stable/#chap-meta) attributes are optional, but it’s still a good idea to provide at least the `description`, `homepage` and [`license`](https://nixos.org/manual/nixpkgs/stable/#sec-meta-license).

Reviewer 2: I'm hesitant to characterize `license` as optional.

>    - All other [`meta`](https://nixos.org/manual/nixpkgs/stable/#chap-meta) attributes are optional, but it’s still a good idea to provide at least the `description`, `homepage` and [`license`](https://nixos.org/manual/nixpkgs/stable/#sec-meta-license).

Reviewer 3: platforms and mainProgram should be set, too.

Author: Personally I'm also expecting the license to be set, but let's have that conversation separately ok?

Reviewer: Requiring license is probably defensible from «checking it is supposed to be packageable» point of view.

Reviewer 8: This document is still a bit of a mess, there's also this section on meta which is probably a better place to change: https://github.com/NixOS/nixpkgs/blob/c5d067d0f23a3412312f036a315bd3e2a8970e9b/pkgs/README.md#meta-attributes

Author: I added the license question in the filters: d7b1941ca8c50ff500474bb864b634e921b41dd7

Agreed, once this is merged, I think we should re-format the doc with a focus on the contributor experience, and how they access that information.
